engine/src/test/java/org/hibernate/validator/test/internal/engine/cascaded/NestedCascadedConstraintsTest.java

```diff
@@ -0,0 +1,234 @@
+						.property( "name" )
+		);
+	}
+
+	private static class EmailAddressMap {
+
+		private final Map<String, List<@Valid EmailAddress>> map = new LinkedHashMap<>();
+
+		private static EmailAddressMap validEmailAddressMap() {
+			List<EmailAddress> validEmailAddresses = Arrays.asList( new EmailAddress( "valid-email-1@example.com" ), new EmailAddress( "valid-email-2@example.com" ) );
+
+			EmailAddressMap emailAddressMap = new EmailAddressMap();
+			emailAddressMap.map.put( "valid", validEmailAddresses );
+
+			return emailAddressMap;
+		}
+
+		private static EmailAddressMap invalidEmailAddressMap() {
+			List<EmailAddress> validEmailAddresses = Arrays.asList( new EmailAddress( "valid-email-1@example.com" ), new EmailAddress( "valid-email-2@example.com" ) );
+			List<EmailAddress> emailAddressesContainingInvalidEmails = Arrays.asList( new EmailAddress( "valid-email-3@example.com" ),
+					new EmailAddress( "invalid-1" ), new EmailAddress( "invalid-2" ) );
+
+			EmailAddressMap emailAddressMap = new EmailAddressMap();
+			emailAddressMap.map.put( "valid", validEmailAddresses );
+			emailAddressMap.map.put( "invalid", emailAddressesContainingInvalidEmails );
+
+			return emailAddressMap;
+		}
+	}
+
+	private static class CinemaEmailAddresses {
+
+		private final Map<@NotNull Optional<@Valid Cinema>, List<@NotNull @Valid EmailAddress>> map = new LinkedHashMap<>();
+
+		private static CinemaEmailAddresses validCinemaEmailAddresses() {
+			CinemaEmailAddresses cinemaEmailAddresses = new CinemaEmailAddresses();
+			cinemaEmailAddresses.map.put(
+					Optional.of( new Cinema( "cinema1", new SomeReference<>( new Visitor( "Name 1" ) ) ) ),
+					Arrays.asList( new EmailAddress( "valid-email-1@example.com" ), new EmailAddress( "valid-email-2@example.com" ) )
+			);
+
+			return cinemaEmailAddresses;
+		}
+
+		private static CinemaEmailAddresses invalidCinemaEmailAddresses() {
+			CinemaEmailAddresses cinemaEmailAddresses = validCinemaEmailAddresses();
+			cinemaEmailAddresses.map.put(
+					Optional.of( new Cinema( "cinema2", new SomeReference<>( new Visitor( "Name 2" ) ) ) ),
+					Arrays.asList( new EmailAddress( "valid-email-3@example.com" ), new EmailAddress( "invalid-1" ), new EmailAddress( "invalid-2" ) )
+			);
+			cinemaEmailAddresses.map.put(
+					Optional.of( new Cinema( "cinema3", new SomeReference<>( new Visitor( "Name 3" ) ) ) ),
+					Arrays.asList( (EmailAddress) null )
+			);
+
+			return cinemaEmailAddresses;
+		}
+
+		private static CinemaEmailAddresses invalidKey() {
+			CinemaEmailAddresses cinemaEmailAddresses = validCinemaEmailAddresses();
+			cinemaEmailAddresses.map.put(
+					Optional.of( new Cinema( "cinema4", new SomeReference<>( new Visitor() ) ) ),
+					Arrays.asList( new EmailAddress( "valid-email-4@example.com" ) )
+			);
+
+			return cinemaEmailAddresses;
+		}
+	}
+
+	@SuppressWarnings({ "unused", "unchecked" })
+	private static class CinemaArray {
+
+		private Optional<@NotNull @Valid Cinema>[] array;
```

Comment:
Using `@NotNull` within `Optional` is weird. Also it's for the sake of the example only, I'd prefer if we found a more realistic usage.